crates/vm-superio/src/serial.rs

```diff
@@ -282,21 +326,68 @@ impl<T: Trigger, EV: SerialEvents, W: Write> Serial<T, EV, W> {
     ///           is not of interest,
     ///           [std::io::Sink](https://doc.rust-lang.org/std/io/struct.Sink.html)
     ///           can be used here.
-    pub fn with_events(trigger: T, serial_evts: EV, out: W) -> Self {
-        Serial {
-            baud_divisor_low: DEFAULT_BAUD_DIVISOR_LOW,
-            baud_divisor_high: DEFAULT_BAUD_DIVISOR_HIGH,
-            interrupt_enable: DEFAULT_INTERRUPT_ENABLE,
-            interrupt_identification: DEFAULT_INTERRUPT_IDENTIFICATION,
-            line_control: DEFAULT_LINE_CONTROL,
-            line_status: DEFAULT_LINE_STATUS,
-            modem_control: DEFAULT_MODEM_CONTROL,
-            modem_status: DEFAULT_MODEM_STATUS,
-            scratch: DEFAULT_SCRATCH,
-            in_buffer: VecDeque::new(),
+    /// * `state` - A reference to the state from which the `Serial` is constructed.
+    pub fn from_state(trigger: T, serial_evts: EV, out: W, state: &SerialState) -> Self {
+        let mut buffer = state.in_buffer.clone();
+        buffer.truncate(FIFO_SIZE);
+
+        let mut serial = Serial {
+            baud_divisor_low: state.baud_divisor_low,
+            baud_divisor_high: state.baud_divisor_high,
+            interrupt_enable: state.interrupt_enable,
+            interrupt_identification: state.interrupt_identification,
+            line_control: state.line_control,
+            line_status: state.line_status,
+            modem_control: state.modem_control,
+            modem_status: state.modem_status,
+            scratch: state.scratch,
+            in_buffer: VecDeque::from(buffer),
             interrupt_evt: trigger,
             events: serial_evts,
             out,
+        };
+
+        if serial.is_thr_interrupt_enabled() && serial.is_thr_interrupt_set() {
+            serial.trigger_interrupt().unwrap_or_default();
```

Comment:
Maybe a small comment why it's safe to unwrap and ignore?

Reply:
That's was my initial instinct as well, but I wasn't sure what's more desirable  for `Serial::from_state()`: to fail fast or cope with programming errors. Thanks for the feedback. Will update the PR so that `from_state()` returns errors instead of being best effort.